### PR TITLE
DS-2552: Makes configurable whether handles or DOIs should be displayed in JSPUI

### PR DESCRIPTION
**Pull request by pnbecker**
Adds the configuration property webui.preferred.identifier and uses it in JSPUI item view.

https://jira.duraspace.org/browse/DS-2552

### CODE REVIEW

**Comment by johlton** [comment]
This is a great improvement! Thanks, Mate!

**Comment by pnbecker** [comment]
Rebased and again ready for review.

**Comment by pnbecker** [comment]
Rebased again.

**Comment by KevinVdV** [comment]
Took a look at the code & noticed that a lot of changes involve calling the DOI code event if this feature isn't enabled. For example in the "HandleServlet" class the DOI variable is retrieved but only used when this feature is enabled. I would only contact the DOI service if this feature is enabled.

The config also doesn't mention that this feature is JSPUI only (at the moment).

File: dspace-jspui/src/main/java/org/dspace/app/webui/components/VersioningItemHome.java [hunk 1]
```diff
@@ -91,26 +99,49 @@ public void process(Context context, HttpServletRequest request,
...
+                                            latestVersionDOI = 
+                                                    doiService.DOIToExternalForm(
+                                                            identifierService.lookup(
+                                                                    context, latestVersionItem, DOI.class));
+                                        }
+                                        catch (IdentifierException ex)
+                                        {
+                                            log.info("Unable to convert DOI '" + latestVersionDOI 
+                                                    + "' into external form: " + ex.toString(), ex);
+                                            throw new PluginException(ex);
```
Comment (KevinVdV) [inline dspace-jspui/src/main/java/org/dspace/app/webui/components/VersioningItemHome.java]: Would throwing an exception here not a bit harsh, perhaps a fallback to handle is more gracious ?

File: dspace-jspui/src/main/java/org/dspace/app/webui/components/VersioningItemHome.java [hunk 2]
```diff
@@ -91,26 +99,49 @@ public void process(Context context, HttpServletRequest request,
...
+                                        String latestVersionDOI = null;
+                                        try {
+                                            latestVersionDOI = 
+                                                    doiService.DOIToExternalForm(
+                                                            identifierService.lookup(
+                                                                    context, latestVersionItem, DOI.class));
+                                        }
+                                        catch (IdentifierException ex)
+                                        {
+                                            log.info("Unable to convert DOI '" + latestVersionDOI
```
Comment (KevinVdV) [inline dspace-jspui/src/main/java/org/dspace/app/webui/components/VersioningItemHome.java]: Should this not be log.error ?

File: dspace-jspui/src/main/java/org/dspace/app/webui/util/UIUtil.java [hunk 3]
```diff
@@ -184,6 +195,78 @@ public static Context obtainContext(HttpServletRequest request)
...
+        String versionHandle = item.getHandle();
+
+        // lookup the version doi
+        String versionDOI = null;
+        try {
+            versionDOI = identifierService.lookup(ctx, item, DOI.class);
+        }
+        catch (Exception ex)
+        {
+            log.info("Unable to lookup DOI for Item " + item.getID()
```
Comment (KevinVdV) [inline dspace-jspui/src/main/java/org/dspace/app/webui/util/UIUtil.java]: Should this not be log.error ?

File: dspace-jspui/src/main/java/org/dspace/app/webui/util/UIUtil.java [hunk 4]
```diff
@@ -184,6 +195,78 @@ public static Context obtainContext(HttpServletRequest request)
...
+            try {
+                dois = new String[] {
+                    handleService.resolveToURL(ctx, versionHandle),
+                    doiService.DOIToExternalForm(versionDOI),
+                    doiService.formatIdentifier(versionDOI).substring(DOI.SCHEME.length()),
+                    doiService.formatIdentifier(versionDOI)
+                };
+            } catch (Exception ex) {
+                dois = null;
+                log.info("Unable to format DOI " + versionDOI
```
Comment (KevinVdV) [inline dspace-jspui/src/main/java/org/dspace/app/webui/util/UIUtil.java]: Should this not be log.error ?

File: dspace-jspui/src/main/java/org/dspace/app/webui/servlet/HandleServlet.java [hunk 5]
```diff
@@ -439,6 +449,38 @@ private void displayItem(Context context, HttpServletRequest request,
...
+        if (doi != null)
+        {
+            try
+            {
+                doi = doiService.DOIToExternalForm(doi);
+            }
+            catch (Exception ex)
+            {
+                doi = null;
+                log.warn("Unable to convert DOI '" + doi + "' into external form.", ex);
```
Comment (KevinVdV) [inline dspace-jspui/src/main/java/org/dspace/app/webui/servlet/HandleServlet.java]: Should this not be log.error ?

File: dspace-jspui/src/main/java/org/dspace/app/webui/components/VersioningItemHome.java (hunk 1 above)
Comment (pnbecker) [inline dspace-jspui/src/main/java/org/dspace/app/webui/components/VersioningItemHome.java]: Yes and no. ;-)
The lines 118 contains two method calls. If I seperate them and call doiService.DOIToExternalForm(...) only if identifierService.lookup(...) returns not null then an exception is appropriate. I'll extend the commit in such a way.

File: dspace-jspui/src/main/java/org/dspace/app/webui/components/VersioningItemHome.java (hunk 2 above)
Comment (pnbecker) [inline dspace-jspui/src/main/java/org/dspace/app/webui/components/VersioningItemHome.java]: Good point, will change this.

File: dspace-jspui/src/main/java/org/dspace/app/webui/util/UIUtil.java (hunk 3 above)
Comment (pnbecker) [inline dspace-jspui/src/main/java/org/dspace/app/webui/util/UIUtil.java]: I actually noted that we do neither need the try-catch-block nor the logging. iderntifierService.lookup(...) doesn't throw an exception (anymore?).

File: dspace-jspui/src/main/java/org/dspace/app/webui/util/UIUtil.java (hunk 4 above)
Comment (pnbecker) [inline dspace-jspui/src/main/java/org/dspace/app/webui/util/UIUtil.java]: I'll enhance the log level.

File: dspace-jspui/src/main/java/org/dspace/app/webui/servlet/HandleServlet.java (hunk 5 above)
Comment (pnbecker) [inline dspace-jspui/src/main/java/org/dspace/app/webui/servlet/HandleServlet.java]: I'll change it.

**Comment by pnbecker** [comment]
@KevinVdV Thanks for looking on this. I updated the PR accordingly to your suggestions.

**Comment by pnbecker** [comment]
resolved merge conflict.

**Comment by tdonohue** [comment]
+1 code looks reasonable to me. If there's any JSPUI users who can give this a quick test, it seems ready to merge.

**Comment by pnbecker** [comment]
Rebased and resolved simple merge conflict.

**Comment by hardyoyo** [comment]
+1 by inspection

**Comment by mwoodiupui** [comment]
Sorry, this has drifted again w.r.t. master.  Would you rebase it one more time, please, and then merge it.  This was discussed on \#dspace today, plus it now has enough votes to go in.

**Comment by pnbecker** [comment]
rebased, but not retested.

**Comment by pnbecker** [comment]
As disscussed in yesterday's developer meeting I tested this again. As all seems to be fine and the code was inspected several times I merged this. I enhanced the comments in dspace.cfg before merging.

**Pull request merged**